fedimint-api/src/db/notifications.rs

```diff
@@ -0,0 +1,82 @@
+use std::{
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
+};
+
+use tokio::sync::Notify;
+
+// arbitrary number. if you have performance problems, tweak this
+const NOTIFY_BUCKETS: usize = 32;
+
+/// Manages notifications for the database
+#[derive(Debug)]
+pub struct Notifications {
+    // TODO: Vec?
+    slots: [Notify; NOTIFY_BUCKETS],
+}
+
+impl Default for Notifications {
+    fn default() -> Self {
+        Self {
+            slots: std::array::from_fn(|_| Notify::new()),
+        }
+    }
+}
+
+impl Notifications {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Waits untils the key *might* have changed.
+    ///
+    /// This registers for notification when called.
+    /// Then waits for the notification when .awaited.
+    pub async fn wait<K>(&self, key: &K)
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.slots.len() as u64)) as usize;
+        // NOTE: this should be the first await point in this function
+        self.slots[slot_idx].notified().await;
+    }
+
+    fn notify_hash(&self, hash_value: u64) {
+        let slot_idx = (hash_value % (self.slots.len() as u64)) as usize;
+        self.slots[slot_idx].notify_waiters();
+    }
+
+    /// Sends the notifications in the queue.
+    ///
+    /// Clears the NotifyQueue.
+    // TODO: should we remove async?
+    pub async fn submit_queue(&self, queue: &mut NotifyQueue) {
+        for hash_value in queue.hashes.drain(..) {
+            self.notify_hash(hash_value);
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct NotifyQueue {
+    hashes: Vec<u64>,
```

Comment:
This could be a `*Set<u64>` imo. Or even just a bit vector (we depend on `bitvec` already anyway).